From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix MC-299642

Related MC issue: https://bugs.mojang.com/browse/MC/issues/MC-299642

Minecraft uses 3D movement control for turtles to move in water, which makes them fall even NoGravity tag is set to true.
This direct manipulation of deltaMovement bypasses the standard move logic where the NoGravity tag is checked and applied.

diff --git a/net/minecraft/world/entity/animal/turtle/Turtle.java b/net/minecraft/world/entity/animal/turtle/Turtle.java
index a3f2076e6f0e51ac51a8af94cae91baf82c247a1..3d4605bcd995ad38769884b6ab422aedbf1a7274 100644
--- a/net/minecraft/world/entity/animal/turtle/Turtle.java
+++ b/net/minecraft/world/entity/animal/turtle/Turtle.java
@@ -574,9 +574,15 @@ public class Turtle extends Animal {
                     float f = (float)(Mth.atan2(d2, d) * 180.0F / (float)Math.PI) - 90.0F;
                     this.turtle.setYRot(this.rotlerp(this.turtle.getYRot(), f, 90.0F));
                     this.turtle.yBodyRot = this.turtle.getYRot();
+                    // Leaf start - Fix MC-299642
+                    if (!this.turtle.isNoGravity() || this.turtle.isInWater()) {
                     float f1 = (float)(this.getSpeedModifier() * this.turtle.getAttributeValue(Attributes.MOVEMENT_SPEED)); // Purpur - Ridables
                     this.turtle.setSpeed(Mth.lerp(0.125F, this.turtle.getSpeed(), f1));
                     this.turtle.setDeltaMovement(this.turtle.getDeltaMovement().add(0.0, this.turtle.getSpeed() * d1 * 0.1, 0.0));
+                    } else {
+                        this.turtle.setSpeed(0.0F);
+                    }
+                    // Leaf end - Fix MC-299642
                 }
             } else {
                 this.turtle.setSpeed(0.0F);
